external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala

```diff
@@ -40,81 +90,114 @@ private[kafka010] object CachedKafkaProducer extends Logging {
-      .build[Seq[(String, Object)], Producer](cacheLoader)
-
-  private def createKafkaProducer(producerConfiguration: ju.Map[String, Object]): Producer = {
-    val updatedKafkaProducerConfiguration =
-      KafkaConfigUpdater("executor", producerConfiguration.asScala.toMap)
-        .setAuthenticationConfigIfNeeded()
-        .build()
-    val kafkaProducer: Producer = new Producer(updatedKafkaProducerConfiguration)
-    logDebug(s"Created a new instance of KafkaProducer for $updatedKafkaProducerConfiguration.")
-    kafkaProducer
-  }
+      .build[Seq[(String, Object)], CachedKafkaProducer](cacheLoader)
 
   /**
    * Get a cached KafkaProducer for a given configuration. If matching KafkaProducer doesn't
    * exist, a new KafkaProducer will be created. KafkaProducer is thread safe, it is best to keep
    * one instance per specified kafkaParams.
    */
-  private[kafka010] def getOrCreate(kafkaParams: ju.Map[String, Object]): Producer = {
-    val paramsSeq: Seq[(String, Object)] = paramsToSeq(kafkaParams)
+  private[kafka010] def acquire(kafkaParamsMap: ju.Map[String, Object]): CachedKafkaProducer = {
+    val paramsSeq: Seq[(String, Object)] = paramsToSeq(kafkaParamsMap)
     try {
-      guavaCache.get(paramsSeq)
+      val producer = this.synchronized {
+        val cachedKafkaProducer: CachedKafkaProducer = guavaCache.get(paramsSeq)
+        cachedKafkaProducer.inUseCount.incrementAndGet()
+        logDebug(s"Granted producer $cachedKafkaProducer")
+        cachedKafkaProducer
+      }
+      producer
     } catch {
-      case e @ (_: ExecutionException | _: UncheckedExecutionException | _: ExecutionError)
+      case e@(_: ExecutionException | _: UncheckedExecutionException | _: ExecutionError)
         if e.getCause != null =>
         throw e.getCause
     }
   }
 
-  private def paramsToSeq(kafkaParams: ju.Map[String, Object]): Seq[(String, Object)] = {
-    val paramsSeq: Seq[(String, Object)] = kafkaParams.asScala.toSeq.sortBy(x => x._1)
+  private def paramsToSeq(kafkaParamsMap: ju.Map[String, Object]): Seq[(String, Object)] = {
+    val paramsSeq: Seq[(String, Object)] = kafkaParamsMap.asScala.toSeq.sortBy(x => x._1)
     paramsSeq
   }
 
-  /** For explicitly closing kafka producer */
-  private[kafka010] def close(kafkaParams: ju.Map[String, Object]): Unit = {
-    val paramsSeq = paramsToSeq(kafkaParams)
-    guavaCache.invalidate(paramsSeq)
+  /* Release a kafka producer back to the kafka cache. We simply decrement it's inuse count. */
+  private[kafka010] def release(producer: CachedKafkaProducer, failing: Boolean): Unit = {
+    this.synchronized {
+      // It should be ok to call release multiple times on the same producer object.
+      if (producer.inUse()) {
+        // So that we do not end up with -ve in-use counts.
+        producer.inUseCount.decrementAndGet()
+        logDebug(s"Released producer $producer.")
+      } else {
+        logWarning(s"Tried to release a not in use producer, $producer.")
```

Comment:
I have left this warning, for testing.

Reply:
I meant this would indicate a bug in logic if my understanding is right (so should not happen):  then worth to log as error or even throw IllegalStateException.

Reply:
Such case exists in consumer side where I've found some minor problems. I've fixed this when I've adapted the related PR to DStreams but this minor issue maybe still exists in structured area. Just like these 2 cases I think the same here, namely if such case happens it's a problem but this shouldn't stop the processing with an exception. In order to conform with consumer side it's good to write this with a warning.